### PR TITLE
Implement open_ncml function converting a NcML document to an xr.Dataset

### PR DESCRIPTION
**Pull request by huard**
Add `open_ncml` function converts NcML document to xarray.Dataset. 

- [x] Include documentation when adding new features.
- [x] Include new tests or update existing tests when applicable.

Auto-generated a *data model* for NcML elements using `xsdata`. This is used to parse the NcML into Python objects (see `generated` directory).
The code then goes through these elements and translate them into xarray instructions. See `parser.py`.
The full NcML syntax is not fully implemented at the moment, but the core elements are there (at least those that I use). 

Note that the test data comes from https://github.com/Unidata/netcdf-java, whose license is BSD-3. Please advise on how to comply with the Apache license used here.

### CODE REVIEW

File: xncml/parser.py [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+
+    # Need to decode time variable
+    if obj.time_units_change:
+        for i, ds in enumerate(items):
+            t = xr.as_variable(ds[obj.dim_name], obj.dim_name)  # Maybe not the same name...
+            encoded = CFDatetimeCoder(use_cftime=True).decode(t, name=t.name)
+            items[i] = ds.assign_coords({obj.dim_name: encoded})
+
+    # Translate different types of aggregation into xarray instructions.
+    match obj.type:
```
Comment (aulemahal) [inline xncml/parser.py:208]: As discussed earlier, it's cool to use the `match` structure, but it is new in python 3.10  and the current numpy/xarray is still accepting python 3.8.
According to NEP 29, Python 3.10+ is scheduled for Apr 5, 2024 ;).

File: xncml/parser.py [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+      Dataset holding variables and attributes defined in <aggregation> element.
+    """
+    from xarray.coding.times import CFDatetimeCoder
+
+    # Names of variables to be aggregated. All variables if undefined.
+    names = [v.name for v in filter_by_class(obj.choice, Aggregation.VariableAgg)]
+
+    # Create list of items to aggregate.
+    items = []
+    for item in obj.netcdf:
```
[the remark this answers is not in the document]
Comment (aulemahal) [reply xncml/parser.py:169]: I may have talked too fast, if this here is the only instance where there's a disconnect, it's not that bad.

File: xncml/parser.py [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+## TODO
+
+Support for these elements is missing:
+- <cacheVariable>
+- <logicalReduce>
+- <logicalSection>
+- <logicalSlice>
+- <promoteGlobalAttribute>
+- <enumTypedef>
+- <group>
```
Comment (aulemahal) [inline xncml/parser.py:26]: what is `read_group` then?

File: xncml/parser.py [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+        path = ncml.parent / path
+
+    files = path.rglob("*") if obj.subdirs else path.glob("*")
+
+    if obj.reg_exp:
+        pat = re.compile(obj.reg_exp)
+        files = filter(pat.match, map(str, files))
+    elif obj.suffix:
+        pat = "*" + obj.suffix
+        files = [f for f in files if f.match(pat)]
```
Comment (aulemahal) [inline xncml/parser.py:317]: To have clean code as above, one could do:
```
import glob  # this is in the standard lib

glob.fnmatch.filter(files, pat)
```
:nerd_face:

File: xncml/parser.py [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        # A JOIN_NEW aggregation has exactly one coordinate value
+        case AggregationType.JOIN_NEW:
+            coord = val
+        case AggregationType.JOIN_EXISTING:
+            if "," in val:
+                coord = val.split(",")
+            elif " " in val:
+                coord = val.split(" ")
+            else:
+                coord = [val]
```
Comment (aulemahal) [inline xncml/parser.py:358]: The annotated schema linked above says : "blanks **and/or** commas", so you might want to do:
```suggestion
            coord = val.replace(',', ' ').split()
```
Split with no arguments catches all whitespaceS and line breaks and returns a list even if there is a single value.

I think my suggestion is slightly less explicit thant yours, but it think catches more corner cases.

File: xncml/parser.py [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+    value = obj.value or obj.content
+    if value:
+        if obj.type in [DataType.STRING, DataType.STRING_1]:
+            return value
+
+        sep = obj.separator or " "
+        values = value.split(sep)
+        return tuple(map(nctype(obj), values))
+
+        return nctype(obj)(value)
```
Comment (aulemahal) [inline xncml/parser.py:577]: Unreachable return :boom: !
```suggestion
```

**Review by aulemahal — COMMENTED** [review]
A first pass! I tried to read the annotated schema in parallel to understand what's going on, but my experience with ncmls is small, so I might have missed things.

Over all, I find this implementation clear and readable!

**Comment by dblodgett-usgs** [comment]
What’s the status of this PR? @huard are you out of cycles to respond to the comments? I may have some team mates who could commit time to help get this going if there were a way to plug in.

**Comment by huard** [comment]
Hi @aulemahal @dblodgett-usgs 

Notifications for this review went into my spam... I'll be able to respond to review comments by tomorrow. 

One thing that bothers me with this PR is that it's using `xsdata`, while xncml uses `xmltodict`. It feels weird to rely on two diffefrent xml libraries in the same package. Suggestions welcome.

File: xncml/parser.py (hunk 3 above)
Comment (huard) [reply xncml/parser.py:26]: Removed from TODO.

File: xncml/parser.py [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+    target = xr.Dataset()
+
+    # <explicit/> element means that only content specifically mentioned in NcML document is included in dataset.
+    if obj.read_metadata is not None:
+        target = ref
+    elif obj.explicit is not None:
+        pass
+    else:  # Default
+        target = ref
+
```
[the remark this answers is not in the document]
Comment (huard) [reply xncml/parser.py:111]: I think you're right.

File: xncml/parser.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (huard) [reply xncml/parser.py:169]: Will fix this.

File: xncml/parser.py (hunk 1 above)
Comment (huard) [reply xncml/parser.py:208]: Will fix.

File: xncml/parser.py (hunk 4 above)
Comment (huard) [reply xncml/parser.py:317]: Nice ! Will do.

**Comment by aulemahal** [comment]
Well, `xsdata` seems to offer more functionality (that you are using in this PR) than `xmltodict`, so I would chose that one? 

Both have no dependencies, so that's not an issue.

**Pull request merged**